File: ImageKleaner.py
import tkinter as tk
import ttkbootstrap as ttk
import tkinter.filedialog, tkinter.messagebox
import random, os
from PIL import Image, ImageOps


class ImageKleaner(tk.Frame):
    def __init__(self, parent, *args, **kwargs):
        tk.Frame.__init__(self, parent, *args, **kwargs)
        self.parent = parent
        self.image_files = [] # list to store the image files path

        self.select_files_button = ttk.Button(self, text="Select image files", command=self.select_files_button_function)
        self.select_files_button.config(width=45)
        self.select_files_button.pack(padx=6, pady=6)

        self.clean_files_button = ttk.Button(self, text="Run image kleaner", command=self.clean_files_button_function)
        self.clean_files_button.config(width=45)
        self.clean_files_button.pack(padx=6, pady=6)

        self.listbox = tk.Listbox(self, height=14)                                            # Listbox widget to display the selected image files
        self.vscrollbar = tk.Scrollbar(self, orient='vertical', command=self.listbox.yview)   # Scrollbar widget for vertical scrolling
        self.hscrollbar = tk.Scrollbar(self, orient='horizontal', command=self.listbox.xview) # Scrollbar widget for horizontal scrolling
        self.listbox['yscrollcommand'] = self.vscrollbar.set                                  # configure the listbox to use the scrollbars
        self.listbox['xscrollcommand'] = self.hscrollbar.set
        self.vscrollbar.pack(side='right', fill='y')                                          # pack the scrollbars to the right and bottom of the frame
        self.hscrollbar.pack(side='bottom', fill='x')
        self.listbox.pack(fill='both', expand=True)                                           # pack the listbox to fill and expand the remaining space

    def select_files_button_function(self):
        self.image_files = list(tkinter.filedialog.askopenfilenames(filetypes=[("Image files", "*.jpg *.jpeg *.png")]))
        self.listbox.delete(0, tk.END) # delete all the items in the listbox
        for file in self.image_files:  # loop through the image files and insert them to the listbox
            self.listbox.insert(tk.END, file)

    def clean_files_button_function(self):
        cwd = os.getcwd()                           # get the current working directory
        if not self.image_files:                    # check if the image files list is empty
            tkinter.messagebox.showinfo("Info", "Please select the files that you want to clean.")
            return
        random_string = "".join(random.choices("0123456789abcdefghijklmnopqrstuvwxyz", k=5))
        folder_name = f"KleanFiles_{random_string}" # create a folder name using randomly generated string
        folder_path = os.path.join(cwd, folder_name)
        tkinter.messagebox.showwarning("Warning", "This may take some time depending on the number of files selected and the app window may be unresponsive.")
        if not os.path.exists(folder_path):         # check if the folder path does not exist
            os.mkdir(folder_path)
        for file in self.image_files:               # loop through the image files
            image = Image.open(file)
            image_copy = image.copy()
            image_copy = ImageOps.exif_transpose(image_copy)
            image_copy.getexif().clear()            # clear the EXIF data from the image copy
            file_name = os.path.basename(file)      # get the file name from the file path
            # save the image copy to the folder path with the same file name
            image_copy.save(os.path.join(folder_path, file_name))
            image.close()
            image_copy.close()
        tkinter.messagebox.showinfo("Info", f"Task completed, cleaned images are in \"{folder_path}\" folder.")


if __name__ == "__main__":
    root = tk.Tk()
    root.geometry("305x320")
    root.resizable(False, False)       # disable the resizing of the root window
    root.title("ImageKleaner")
    root.tk.call('tk', 'scaling', 1.2) # set the scaling factor of the root window to 1.2
    ImageKleaner(root).pack(side="top", fill="both", expand=True)
    root.mainloop()                    # start the main loop of the root window

File: ImageKleaner.java
import javax.imageio.ImageIO;
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

// Extend JFrame to create a GUI window
public class ImageKleaner extends JFrame {
    // Private fields to store the image files and GUI components
    private List<File> imageFiles;
    private JList<File> listbox;
    private DefaultListModel<File> listModel;
    private JButton selectFilesButton;
    private JButton cleanFilesButton;

    // Constructor that initializes the fields and sets up the GUI
    public ImageKleaner() {
        super("ImageKleaner");                                         // Call the superclass constructor with the window title
        this.imageFiles = new ArrayList<>();                                 // Initialize the image files list as an empty array list
        this.listModel = new DefaultListModel<>();                           // Initialize the list model as a default list model
        this.listbox = new JList<>(listModel);                               // Initialize the list box as a JList with the list model
        this.listbox.setLayoutOrientation(JList.VERTICAL);                   // Set the layout orientation of the list box to vertical
        JScrollPane listScroller = new JScrollPane(listbox);                 // Create a scroll pane to wrap the list box
        listScroller.setPreferredSize(new Dimension(265, 235)); // Set the preferred size of the scroll pane

        // Initialize the select files button as a JButton with the label
        selectFilesButton = new JButton("Select image files");
        // Add an action listener to the button that calls a method when clicked
        selectFilesButton.addActionListener(this::selectFilesButtonFunction);
        // Initialize the clean files button as a JButton with the label
        cleanFilesButton = new JButton("Run image kleaner");
        // Add an action listener to the button that calls a method when clicked
        cleanFilesButton.addActionListener(this::cleanFilesButtonFunction);

        this.setLayout(new FlowLayout());    // Set the layout of the window to a flow layout
        // Add the buttons and the scroll pane to the window
        this.add(selectFilesButton);
        this.add(cleanFilesButton);
        this.add(listScroller);
        this.setSize(305, 320); // Set the size of the window
        // Set the default close operation of the window to exit
        this.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        this.setResizable(false);  // Set the window to be non-resizable
    }

    // Method that handles the action of selecting files button
    private void selectFilesButtonFunction(ActionEvent e) {
        // Create a file chooser object to let the user select multiple image files
        JFileChooser fileChooser = new JFileChooser();
        fileChooser.setMultiSelectionEnabled(true);      // Enable the multi-selection mode of the file chooser
        // Set the file filter to only accept image files with jpg, jpeg, or png extensions
        fileChooser.setFileFilter(new javax.swing.filechooser.FileNameExtensionFilter("Image files", "jpg", "jpeg", "png"));
        int returnVal = fileChooser.showOpenDialog(this);  // Show the file chooser dialog and get the return value
        if (returnVal == JFileChooser.APPROVE_OPTION) {    // If the user approved the selection, proceed to the next step
            File[] files = fileChooser.getSelectedFiles(); // Get the selected files as an array of File objects
            listModel.clear();                             // Clear the list model to remove any previous files
            // Loop through the selected files and add them to the image files list and the list model
            for (File file : files) {
                imageFiles.add(file);
                listModel.addElement(file);
            }
        }
    }

    // Method that handles the action of cleaning files button
    private void cleanFilesButtonFunction(ActionEvent e) {
        // Check if the image files list is empty, if so, show a message to the user and return
        if (imageFiles.isEmpty()) {
            JOptionPane.showMessageDialog(this, "Please select the files that you want to clean.", "Info", JOptionPane.INFORMATION_MESSAGE);
            return;
        }
        String randomString = generateRandomString(5); // Generate a random string of 5 characters to use as a folder name suffix to prevent overwriting
        String folderName = "KleanFiles_" + randomString;     // Concatenate the prefix "KleanFiles_" with the random string to form the folder name
        File folderPath = new File(folderName);               // Create a file object with the folder name as the path
        if (!folderPath.exists()) {                           // Check if the folder path does not exist, if so, create a new folder
            folderPath.mkdir();
        }
        // Loop through the image files and clean them
        for (File file : imageFiles) {
            try {
                // Read the image data from the file as a buffered image object
                BufferedImage image = ImageIO.read(file);

                // Get the file extension from the file name by finding the last dot and getting the substring after it
                String fileName = file.getName();
                String fileExtension = fileName.substring(fileName.lastIndexOf(".") + 1);

                // Create a new file object with the folder path and the file name as the path
                File outputFile = new File(folderPath, fileName);
                // Write the image data to the new file without copying the EXIF data
                // Use the file extension as the output format
                ImageIO.write(image, fileExtension, outputFile);
            } catch (IOException ex) {
                // Print the stack trace of the exception if any error occurs
                ex.printStackTrace();
            }
        }
        // Show a message to the user that the task is completed and the cleaned images are in the folder
        JOptionPane.showMessageDialog(this, "Task completed, cleaned images are in \"" + folderPath.getAbsolutePath() + "\" folder.", "Info", JOptionPane.INFORMATION_MESSAGE);
    }

    // Method that generates a random string of a given length
    private String generateRandomString(int length) {
        // Define a string of characters to choose from
        String characters = "0123456789abcdefghijklmnopqrstuvwxyz";
        // Create a string builder object to append the characters
        StringBuilder result = new StringBuilder();
        // Create a random object to generate random numbers
        Random random = new Random();
        while (length-- > 0) {    // Loop until the length is zero
            // Append a random character from the characters string to the result
            result.append(characters.charAt(random.nextInt(characters.length())));
        }
        return result.toString(); // Return the result as a string
    }

    // Define the main method that runs the program
    public static void main(String[] args) {
        // Invoke the event queue to run the GUI in a separate thread
        EventQueue.invokeLater(() -> {
            // Create an instance of the ImageKleaner class
            ImageKleaner frame = new ImageKleaner();
            // Set the frame to be visible
            frame.setVisible(true);
        });
    }
}

